core/sched/rt-rm.c

```diff
@@ -0,0 +1,233 @@
+    /* Set entry_data_rm's fields */
+
+    LIST_DELINIT(&ed->head);
+
+    return 0;
+}
+
+/**
+ * Main scheduler routine in RT - Rate Monotonic policy implementation
+ */
+int sched_rm_do_schedule(struct scheduler *s, uint64_t *expiration)
+{
+    struct sched_data_rm *sd = (struct sched_data_rm *) (s->sd);
+
+    struct entry_data_rm *current_ed = NULL;
+    struct entry_data_rm *next_ed = NULL;
+
+    // bool is_switching_needed = false;
+    int next_vcpuid = VCPUID_INVALID;
+
+    sd->tick_count += 1;
+
+    if (LIST_IS_EMPTY(&sd->runqueue)) {
+        // printf("Nothing to run\n");
+        // printf("Idle mode not implemented\n");
+        while(1);
+    }
+
+    /* Check & decrease all entries' period count */
+    struct entry_data_rm *ed = NULL;
+    if (true) { // if (is_switching_needed) {
+        uint32_t min_period = 0xFFFFFFFF;
+        next_ed = NULL;
+
+        LIST_FOR_EACH_ENTRY(ed, &sd->runqueue, head) {
+            if (ed->period_cntdn == 0) {
+                ed->budget_cntdn = ed->budget;
+                ed->period_cntdn = ed->period;
+            }
+
+            /* find the entry with budget left and shortest period */
+            if (ed->budget_cntdn > 0 && min_period > ed->period) {
+                next_ed = ed;
+                min_period = ed->period;
+            }
+
+            ed->period_cntdn -= 1;
+        }
+
+        if (sd->current != NULL) {
+            current_ed = sd->current;
+
+            if (current_ed->budget_cntdn == 0) { // out of budget
+                current_ed->e->state = SCHED_WAITING;
+            } else if (current_ed->period > next_ed->period) { // preemption!
+#ifdef SCHED_RM_REPORT
+                current_ed->preempted += 1;
+#endif
+            } else { // continue to run
+                next_ed = current_ed;
+            }
+        }
+
+        if (next_ed != NULL) {
+            struct sched_entry *next_e = NULL;
+            next_e = next_ed->e;
+            next_e->state = SCHED_RUNNING;
+
+            sd->current = next_ed;
+            next_ed->budget_cntdn -= 1;
+
+#ifdef SCHED_RM_REPORT
+            next_ed->foreground_ticks += 1;
+#endif
+
+            next_vcpuid = next_e->vcpuid;
+        } else {
+            /* TODO:(igkang) handle the situation that there is no entry to run */
+            // printf("Nothing to run\n");
+            // printf("Idle mode not implemented\n");
+            while(1);
+        }
+    }
+
+#ifdef SCHED_RM_REPORT
+    if (sd->tick_count >= sd->tick_1kcount) {
+        printf("RM report (for %u ticks):\n", sd->tick_interval_us);
+
+        LIST_FOR_EACH_ENTRY(ed, &sd->runqueue, head) {
+            printf("    vcpu%u: fg=%u preempt=%u\n", ed->e->vcpuid, ed->foreground_ticks, ed->preempted);
+            ed->foreground_ticks = 0;
+            ed->preempted = 0;
+        }
+
+        sd->tick_1kcount += 1000;
+    }
+#endif
+
+    if (*expiration == 0) {
+        *expiration = timer_get_timenow();
```

Comment:
We already discussed the prefix of function but, in this case, get_timenow is better, and get_current_time is more better.